lib/Platform/Intl/impl_icu/Collator.cpp

```diff
@@ -0,0 +1,592 @@
+  // "best fit"; the default is "best fit". For information about this option,
+  // see the Intl page.
+  auto matcherRes = OptionHelpers::getStringOption(
+      runtime,
+      options,
+      constants::opt_name::localeMatcher,
+      constants::opt_value::locale_matcher::validLocaleMatchers,
+      constants::opt_value::locale_matcher::best_fit);
+  if (LLVM_UNLIKELY(matcherRes == vm::ExecutionStatus::EXCEPTION)) {
+    return vm::ExecutionStatus::EXCEPTION;
+  }
+  // 9. Set opt.[[localeMatcher]] to matcher.
+  opt.emplace(constants::opt_name::localeMatcher, **matcherRes);
+
+  // 10. Let collation be ? GetOption(options, "collation", "string", empty,
+  // undefined).
+  auto collationValueRes = OptionHelpers::getStringOption(
+      runtime, options, constants::opt_name::collation, {}, std::nullopt);
+  if (LLVM_UNLIKELY(collationValueRes == vm::ExecutionStatus::EXCEPTION)) {
+    return vm::ExecutionStatus::EXCEPTION;
+  }
+  // 11. If collation is not undefined, then
+  // a. If collation does not match the Unicode Locale Identifier type
+  // nonterminal, throw a RangeError exception.
+  if (collationValueRes->has_value()) {
+    if (!isUnicodeExtensionType(**collationValueRes)) {
+      return runtime.raiseRangeError(
+          vm::TwineChar16("Invalid collation: ") +
+          vm::TwineChar16((*collationValueRes)->c_str()));
+    }
+    // 12. Set opt.[[co]] to collation.
+    opt.emplace(constants::extension_key::co, **collationValueRes);
+  }
+
+  // 13. Let numeric be ? GetOption(options, "numeric", "boolean", empty,
+  // undefined).
+  auto numericValue = OptionHelpers::getBoolOption(
+      options, constants::opt_name::numeric, std::nullopt);
+  // 14. If numeric is not undefined, then
+  // a. Let numeric be ! ToString(numeric).
+  // Note: We omit the ToString(numeric) operation as it's not observable.
+  // GetBoolOption returns a Boolean and ToString(Boolean) does not
+  // have side effects.
+  // 15. Set opt.[[kn]] to numeric.
+  if (numericValue.has_value()) {
+    opt.emplace(constants::extension_key::kn, *numericValue);
+  }
+
+  // 16. Let caseFirst be ? GetOption(options, "caseFirst", "string", « "upper",
+  // "lower", "false" », undefined).
+  auto caseFirstValueRes = OptionHelpers::getStringOption(
+      runtime,
+      options,
+      constants::opt_name::caseFirst,
+      constants::opt_value::case_first::validCaseFirsts,
+      std::nullopt);
+  if (LLVM_UNLIKELY(caseFirstValueRes == vm::ExecutionStatus::EXCEPTION)) {
+    return vm::ExecutionStatus::EXCEPTION;
+  }
+  // 17. Set opt.[[kf]] to caseFirst.
+  if (caseFirstValueRes->has_value()) {
+    opt.emplace(constants::extension_key::kf, **caseFirstValueRes);
+  }
+
+  // The relevant unicode extensions accepted by Collator as specified here:
+  // https://tc39.github.io/ecma402/#sec-intl-collator-internal-slots
+  // 18. Let relevantExtensionKeys be %Collator%.[[RelevantExtensionKeys]].
+  std::unordered_set<std::u16string> relevantExtensionKeys{
```

Comment:
Can this be a constexpr array, instead of constructing a new set every time? (similar to the Apple implementation)

Reply:
Done in 3rd commit.